File: CodingSamples/Java/Platform-Enterprise/ThreadTest/demo2/Program3.java
class Program {
    
    static class JointAccount {

        private int balance;

        public int balance() {
            return balance;
        }

        public boolean withdraw(int amount){
            boolean success = false;
            //in order to execute code within a synchronized block a thread
            //must acquire the ownership of the built-in monitor of the 
            //specified object(this), only one thread can be the owner of a
            //given object's monitor at a time while other threads must wait 
            //for this thread to release the ownership
            synchronized(this){
                if(balance >= amount){
                    balance = Activity.perform(balance, amount, -1);
                    success = true;
                }
            }
            return success;
        }

        public synchronized boolean withdrawAfterDeposit(int amount) throws InterruptedException {
            //release the ownership of this object's monitor, wait for
            //this monitor to be notified and resume execution after
            //reaquring the monitor 
            this.wait();
            return withdraw(amount);
        }

        public synchronized void deposit(int amount) {
            balance = Activity.perform(balance, amount, 1);
            //when the monitor is released it should go to any thread
            //which released the monitor by calling wait
            this.notify();
        }
    }

    public static void main(String[] args) throws Exception {
        var acc = new JointAccount();
        acc.deposit(10000);
        System.out.println("Joint account opened for Jack and Jill.");
        System.out.printf("Initial Balance: %d%n", acc.balance());
        Thread child1 = new Thread(() -> {
            System.out.println("Jack is withdrawing 6000...");
            if(!acc.withdraw(6000))
                System.out.println("Jack's withdraw operation failed!");
        });
        Thread child2 = new Thread(() -> {
            System.out.println("Jill is withdrawing 7000...");
            if(!acc.withdraw(7000))
                System.out.println("Jill's withdraw operation failed!");
        });
        child1.start();
        child2.start();
        child1.join(); //wait for child1 to exit
        child2.join(); //wait for child2 to exit
        System.out.printf("Final Balance  : %d%n", acc.balance());
    }
}

File: CodingSamples/DotNet/Platform-Enterprise/ThreadTest/DemoApp2/Program3.cs
namespace DemoApp;

class Program
{
    class JointAccount
    {
        public int Balance { get; private set; }

        public bool Withdraw(int amount)
        {
            bool success = false;
            //a thread requests ownership of the monitor associated with object obj
            //by calling Monitor.Enter(obj), if this monitor is not presently owned
            //by another thread the caller thread acquires the ownership and resumes
            //otherwise it must wait until the present owner releases the ownership
            Monitor.Enter(this);
            try
            {
                if(Balance >= amount)
                {
                    Balance = Activity.Perform(amount, Balance, Activity.Debit);
                    success = true;
                }
            }
            finally
            {
                //thread which owns the monitor associated with object obj
                //calls Monitor.Exit(obj) to release its ownership
                Monitor.Exit(this);
            }
            return success;
        }

        public void Deposit(int amount)
        {
            lock(this)
            {
                Balance = Activity.Perform(amount, Balance, Activity.Credit);
            }
        }
    }

    static void Main(string[] args)
    {
        var acc = new JointAccount();
        acc.Deposit(10000);
        Console.WriteLine("Joint account opened for Jack and Jill.");
        Console.WriteLine("Initial Balance = {0}", acc.Balance);
        var child1 = new Thread(() => 
        {
            Console.WriteLine("Jack is withdrawing 6000...");
            if(!acc.Withdraw(6000))
                Console.WriteLine("Jack's withdrawal failed!");
        });
        var child2 = new Thread(() => 
        {
            Console.WriteLine("Jill is withdrawing 7000...");
            if(!acc.Withdraw(7000))
                Console.WriteLine("Jill's withdrawal failed!");
        });
        child1.Start();
        child2.Start();
        child1.Join();
        child2.Join();
        Console.WriteLine("Final Balance = {0}", acc.Balance);
    }
}
